### PR TITLE
P2 signup: P2 Complete Profile

### PR DESCRIPTION
**Pull request by lamosty**
Addresses 4741-gh-automattic/p2.

In this PR, we add a new step to the P2 signup flow:

Screenshot on 2022-03-18 at 18-04-35

It allows uploading or modifying the user's Gravatar and changing their full name (`display_name`).

## Testing instructions

With an incognito window, navigate to `/start/p2`. Using a 10 minute email address, create a new user and verify the email. In the first tab, reload the signup flow (as the email confirmation link will take you to wordpress.com/production). After clicking on Continue, you should arrive to the Complete Profile step.

Try to upload a Gravatar and change your full name. The name must be longer than 3 characters and the Gravatar must be an image.

### CODE REVIEW

File: client/blocks/edit-gravatar/index.jsx
Comment (lamosty) [inline client/blocks/edit-gravatar/index.jsx:234]: Had to modify the original `EditGravatar` component to accomplish this.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lezama) [inline client/signup/steps/p2-complete-profile/index.jsx]: missing translation?

**Comment by lezama** [comment]
works great, let's ship this!

@tyxla could you give this a quick 👀 ? 🙏

File: client/blocks/edit-gravatar/index.jsx
Comment (tyxla) [reply client/blocks/edit-gravatar/index.jsx:234]: I wonder if we could have declared those as child components and just used the `children` prop here.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: By all means, let's avoid introducing new Lodash instances. We've been refactoring away from it for a while now, see p4TIVU-9Bf-p2

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Seems like `props.step.errors` will be an array, so `.length` should work instead of using Lodash's `isEmpty()`

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Seems like `fields.fullName` will be an array, so `.length` should work instead of using Lodash's `isEmpty()`

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Is there really a possibility for `props` to be nullish?

What about the `step` prop?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Noting that the artificial limitations on names could be discriminatory. I'm not convinced that one can't have a 2-character full name nowadays.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: If we don't need the number, can we just omit setting a special array key?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Won't this cause memory leaks or weird side effects due to the fact that requests are performed on behalf of a component that has already been unmounted?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Same as for the above constant.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Any good reason not to declare the debounce timeout inline?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Can be shortened to:

```suggestion
			initialState,
```

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: This state property doesn't seem to be used.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: This state property doesn't seem to be used.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: `this.logValidationErrorToLogstash()` is undefined 🤔  Did we mean to copy it from the site step? Do we actually need this logging? Can we remove it if not?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: I find it a bit confusing that `initialState` is being mutated and has so much logic in the first place. Usually, initial state is static and if it changes during the component's lifecycle, it's practically no longer "initial".

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: I find the entire validation logic here extremely complex just for the sake of validating a single text field. I bet we can write the validation in a much simpler way and much less code without using `formState`. Have you tried that? I'd very much recommend avoiding it for such simple use cases.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Since `<ValidationFieldset />` is the sole child component of this fragment, we can remove the fragment altogether.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Auto-focusing the fields is not recommended for a good reason. Are we certain we really need it here?

In any case, `={ true }` is redundant:

```suggestion
						autoFocus // eslint-disable-line jsx-a11y/no-autofocus
```

Although I'd vote for removing this prop completely.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: This seems wrong - I don't see a `site` field in this component.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: I don't see a `sanitize` method declared in this component. Perhaps we can just ditch this line?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Can we avoid mutating objects like that? Mutability can cause terrible bugs.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: What happens if this request fails? Are we handling it in any way?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: What happens if this request fails? Are we handling it in any way?

Also do we need to wait for it to finish?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Abstractions like that are why I really recommend not using `formState`. Just for the protocol 😉

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [reply client/signup/steps/p2-complete-profile/index.jsx]: This. Also the string seems unfinished.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Do we need the span? If we need it for styling, can we add a classname and style by that classname?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Can we abstract that callback into a method so it doesn't get redeclared on every re-render?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Another part of the string that is missing translation here.

**Review by tyxla — COMMENTED** [review]
> @tyxla could you give this a quick 👀 ? 🙏

Sure thing @lezama 👍

**Comment by lamosty** [comment]
Thanks for the feedback @tyxla! Sorry for all the troubles, I've mostly copied the `P2Site` (https://github.com/Automattic/wp-calypso/blob/f231e5efac926c6a70acbd0fc8b8b8edacb21357/client/signup/steps/p2-site/index.jsx) which is itself copied from the `Site` components which is quite old. Also, as the form was simplified (we dropped the timezone field), I might try to drop the usage of `formState` and go with a much simpler stateless implementation.

File: client/blocks/edit-gravatar/index.jsx
Comment (lamosty) [reply client/blocks/edit-gravatar/index.jsx:234]: Yeah, not sure. Should it have children, though? I'd say this is more of a special case as what would be the children of `EditGravatar` in the logical sense?

I'll defer to your call.

Comment (tyxla) [reply client/blocks/edit-gravatar/index.jsx:234]: It's just at the end of the component and having it as `children` made sense to me. It also makes the component more generic and less specific, and also, you're using it as a similar syntax as children already, just with a different prop name 😉 . Happy to use `children` for that.

Comment (lamosty) [reply client/blocks/edit-gravatar/index.jsx:234]: Great, will do. 👍

**Comment by lamosty** [comment]
@tyxla refactor done. Could you please take another look?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: Changed to 2 chars.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: Will do.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: Right now, the Calypso's data layer handles it by showing a "There was a problem saving your changes." notice. Not sure how rare it would be and whether we need to intercept it in this step and tell user to try again or something or just continue the flow as we do now.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: `submitSignupStep` is the regular signup facility to continue the flow. I'm not sure if there could be a failure in submitting the step and the other steps just submit it, too. If the API request to update user full name which is called above this fails we could perhaps stop the signup process and tell the user to try again?

**Comment by lamosty** [comment]
After a rebase, the primary button is broken (recent issue). We are working on a fix in another PR. Will rebase afterwards.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [reply client/signup/steps/p2-complete-profile/index.jsx]: IMHO, the probability for it to fail is the same as any other request. The difference is that we do handle some of those failures gracefully, while here, we'd likely fail silently (silently for the user at least).

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [reply client/signup/steps/p2-complete-profile/index.jsx]: Sounds good, any kind of failure handling makes sense when there's a request in place that could possibly fail.

File: client/blocks/edit-gravatar/index.jsx
Comment (tyxla) [inline client/blocks/edit-gravatar/index.jsx:235]: Hm, so we do need `this.onReceiveFile` out here. Guess this wasn't such a good use for `children` after all 😭 Mind reverting this change? I owe you a beer.

**Comment by tyxla** [comment]
> After a rebase, the primary button is broken (recent issue). We are working on a fix in another PR. Will rebase afterwards.

I see much (if not all) of the feedback has been addressed. Let me know when this one is ready for a review and I'll take another look. Thanks!

File: client/blocks/edit-gravatar/index.jsx
Comment (lamosty) [reply client/blocks/edit-gravatar/index.jsx:235]: Hehe, sure. Done. 😅

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: Yep, probability is the same but I guess what I meant is that this is not an important request. Worst case user will not have their name set. The whole step is kind of additional. But yeah, I'll try to intercept the error and stop the submit in that case.

**Comment by lamosty** [comment]
@tyxla I added a simple error handling in [f6b4d1d](https://github.com/Automattic/wp-calypso/pull/61914/commits/f6b4d1d68670acbe3c47c9b0ced8768022e531de) but not sure if it's the way to go. Could you give it another pass, please? I think I've addressed all the other points.

**Comment by annemirasol** [comment]
Good job, Rasto!

I noticed that it makes me go through the Profile step again, even though I have already submitted an avatar and name in a previous signup flow.

1. Go through signup flow and complete your profile. Optionally, create a workspace.
2. Go to `calypso.localhost:3000/start/p2` again, and notice that you get the profile page again.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx:146]: I find it very odd that we use both `i18n-calypso` and `@wordpress/i18n` for translations. Can we use just one of them in a single component?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Any particular reason why we're not listing all dependencies of this hook here?

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/index.jsx]: Seems like we require at least 2 characters but the error says at least 3?

File: client/signup/steps/p2-complete-profile/style.scss
Comment (tyxla) [inline client/signup/steps/p2-complete-profile/style.scss]: We already declare styles for this particular selector at the top of the file. Should we merge both style blocks?

**Review by tyxla — COMMENTED** [review]
Haven't tested, just adding a few more suggestions on the code before we go ahead to merge this.

**Comment by lamosty** [comment]
@annemirasol I've added the logic to skip the step if user has already filled in their display name in [a15467e](https://github.com/Automattic/wp-calypso/pull/61914/commits/a15467e750f31fde94649b379dc95649acc971da). I wanted to add the same logic for the Gravatar but at the moment I don't know how to distinguish between the default Gravatar url and a custom one. The default Gravatar url is different for every user.

File: client/signup/steps/p2-complete-profile/index.jsx
Comment (lamosty) [reply client/signup/steps/p2-complete-profile/index.jsx]: I was thinking to include only those which could actually change and be important to react to their change. The others are just functions and stuff. I'll include them all.

**Comment by lamosty** [comment]
Thanks for the very detailed review @tyxla! Points addressed.

**Review by lezama — APPROVED** [review]
let's 🚢 this! Thanks @tyxla for your 👀 once more!

**Comment by annemirasol** [comment]
> I've added the logic to skip the step if user has already filled in their display name in [a15467e](https://github.com/Automattic/wp-calypso/pull/61914/commits/a15467e750f31fde94649b379dc95649acc971da). I wanted to add the same logic for the Gravatar but at the moment I don't know how to distinguish between the default Gravatar url and a custom one. The default Gravatar url is different for every user.

I think this is okay and may be best, actually. I personally don't want to be bugged more than once about uploading a personal avatar. The name is a different matter, since we tend to see a lot of usernames that are unrelated to the person's name.

Nit: maybe `excludeStepIfProfileComplete`? `details` just seem general.

**Review by tyxla — APPROVED** [review]
Haven't tested, but the code's looking great!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7209197

Thank you @lamosty for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7209197).

**Pull request merged**
